Web Automation: modernize callback thunks for async protocol commands

burg opened the pull request:
#### 9e1b4bc5f110647270eef4dfff5d7dc8e61337a2
<pre>
Web Automation: modernize callback thunks for async protocol commands
<a href="https://bugs.webkit.org/show_bug.cgi?id=284470">https://bugs.webkit.org/show_bug.cgi?id=284470</a>
&lt;<a href="https://rdar.apple.com/141291944">rdar://141291944</a>&gt;

Reviewed by Devin Rousso and Patrick Angle.

To avoid code duplication, we need to be able to connect async `WebAutomationSession` / `WebDriverBidiProcessor`
methods to each other. Right now these callbacks are all `BackendDispatcher::CallbackBase` subclasses,
which does not have a standalone constructor. So, we cannot easily pass a lambda that chains completion handlers.

In order to chain callbacks, switch from using `CallbackBase` subclasses to using `Function&lt;&gt;` with proper type parameters.
We can move away from CallbackBase by capturing any needed state in a closure and changing the type of callbacks to mention return types.
We deploy the new type aliases `CommandResponse&lt;T&gt;` and `CommandCallback&lt;T&gt;` to make code more readable, adding the &apos;Of&apos;
suffix for multiple arguments.

```
namespace Inspector {
    typedef String ErrorString;

    template &lt;typename T&gt;
    using ErrorStringOr = Expected&lt;T, ErrorString&gt;;

    template &lt;typename T&gt;
    using CommandResponse = Inspector::Protocol::ErrorStringOr&lt;T&gt;;

    template &lt;typename T&gt;
    using CommandResponseOf = Inspector::Protocol::ErrorStringOr&lt;std::tuple&lt;T&gt;&gt;;

    template &lt;typename T&gt;
    using CommandCallback = Function&lt;void(CommandResponse&lt;T&gt;)&gt;;

    template &lt;typename T&gt;
    using CommandCallbackOf = Function&lt;void(CommandResponseOf&lt;T&gt;)&gt;;
}
```

These changes are scoped to *not* affect existing Web Inspector protocol generated code. I&apos;d love to remove
uses of `CallbackBase` from the existing uses of &apos;async&apos; commands entirely. However, doing so is essentially the
same work as refactoring/modernizing IndexedDBAgent. We can do that at a time of our choosing. Fixing this is
tracked by &lt;<a href="https://webkit.org/b/288203">https://webkit.org/b/288203</a>&gt;.

The remainder of this commit is updating callsites in `WebAutomationSession` to reflect that calling `sendSuccess()` /
`sendFailure()` has been superceded by returning expected or unexpected values of `CommandResult&lt;T&gt;` (aka `ErrorStringOr&lt;T&gt;`).
Also, adapt code to use `CommandCallback&lt;T&gt;` for zero and single parameter async commands, and `CommandCallbackOf&lt;T&gt;` for
async commands with more than one return value.

* Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:
Introduce new type aliases which make it easier to express the type of an async callback in command implementations.

* Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py:
Split out the declaration of the reply thunk class from the command&apos;s method declaration.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_header.py:
(CppBackendDispatcherHeaderGenerator._generate_handler_declarations_for_domain):
(CppBackendDispatcherHeaderGenerator._generate_async_response_thunk_declaration_for_command): Added.
(CppBackendDispatcherHeaderGenerator):
(CppBackendDispatcherHeaderGenerator._generate_async_handler_declaration_for_command):
Now that the return types are part of the type signature, enhance `_generate_dispatcher_declaration_for_command`
to compute the type arguments for `Inspector::CommandCallbackOf&lt;T...&gt;`.
Add a new framework setting so that we can continue to generate existing async protocol code
for non-WebDriver domains.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementations_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_async_command_response_thunk_class_for_domain): Added.
(CppBackendDispatcherImplementationGenerator._generate_async_dispatcher_class_for_domain): Deleted.
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementation_for_command):
Refactor `_generate_dispatcher_implementation_for_command` so that the code to convert return parameters
is shared between the sync and async command cases. Move the return parameter passing code from the thunk
class to the anonymous lambda function passed as the last argument. Inside the lambda, capture what would
have been stored in the thunk class&apos;s member variables so that we can associate the error with the
originating command request (`protocol_requestId`).

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::terminate):
(WebKit::WebAutomationSession::getBrowsingContext):
(WebKit::WebAutomationSession::createBrowsingContext):
(WebKit::WebAutomationSession::switchToBrowsingContext):
(WebKit::WebAutomationSession::setWindowFrameOfBrowsingContext):
(WebKit::WebAutomationSession::waitForNavigationToComplete):
Fix this function so that it explicitly returns after moving `callback` into
the queue of callbacks waiting for navigation. It&apos;s not safe to use it anyway,
so simplify the control flow to make it easier to read.

(WebKit::WebAutomationSession::waitForNavigationToCompleteOnPage):
(WebKit::WebAutomationSession::waitForNavigationToCompleteOnFrame):
(WebKit::WebAutomationSession::respondToPendingPageNavigationCallbacksWithTimeout):
(WebKit::WebAutomationSession::respondToPendingFrameNavigationCallbacksWithTimeout):
(WebKit::WebAutomationSession::maximizeWindowOfBrowsingContext):
(WebKit::WebAutomationSession::hideWindowOfBrowsingContext):
(WebKit::WebAutomationSession::willShowJavaScriptDialog):
(WebKit::WebAutomationSession::navigateBrowsingContext):
(WebKit::WebAutomationSession::goBackInBrowsingContext):
(WebKit::WebAutomationSession::goForwardInBrowsingContext):
(WebKit::WebAutomationSession::reloadBrowsingContext):
(WebKit::WebAutomationSession::navigationOccurredForFrame):
(WebKit::WebAutomationSession::documentLoadedForFrame):
(WebKit::WebAutomationSession::inspectorFrontendLoaded):
(WebKit::WebAutomationSession::mouseEventsFlushedForPage):
(WebKit::WebAutomationSession::keyboardEventsFlushedForPage):
(WebKit::WebAutomationSession::wheelEventsFlushedForPage):
(WebKit::WebAutomationSession::willClosePage):
(WebKit::WebAutomationSession::evaluateJavaScriptFunction):
(WebKit::WebAutomationSession::resolveChildFrameHandle):
(WebKit::WebAutomationSession::resolveParentFrameHandle):
(WebKit::WebAutomationSession::computeElementLayout):
(WebKit::WebAutomationSession::getComputedRole):
(WebKit::WebAutomationSession::getComputedLabel):
(WebKit::WebAutomationSession::selectOptionElement):
(WebKit::WebAutomationSession::setFilesForInputFileUpload):
(WebKit::WebAutomationSession::getAllCookies):
(WebKit::WebAutomationSession::deleteSingleCookie):
(WebKit::WebAutomationSession::addSingleCookie):
Add missing move for `callback` to be owned the completion handler.

(WebKit::WebAutomationSession::simulateMouseInteraction):
(WebKit::WebAutomationSession::simulateTouchInteraction):
(WebKit::WebAutomationSession::simulateKeyboardInteraction):
(WebKit::WebAutomationSession::simulateWheelInteraction):
These cases require a bit more care to bridge between `CommandResponse&lt;void&gt;`
and `AutomationCommandError`.

(WebKit::WebAutomationSession::performMouseInteraction):
(WebKit::WebAutomationSession::performKeyboardInteractions):
(WebKit::pressedCharKey):
(WebKit::WebAutomationSession::performInteractionSequence):
(WebKit::WebAutomationSession::cancelInteractionSequence):
(WebKit::WebAutomationSession::takeScreenshot):
Adapt to using `CommandCallback&lt;void&gt;` and `CommandCallback&lt;T&gt;`.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSessionMacros.h:
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::WebAutomationSession::inspectBrowsingContext):
Adapt to using `CommandCallback&lt;void&gt;` and `CommandCallback&lt;T&gt;`.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSessionMacros.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::navigate): Update signature.

Canonical link: <a href="https://commits.webkit.org/291284@main">https://commits.webkit.org/291284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e520bec7aa163cb7f5c381887a4894fc42103cea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92452 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11994 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1596 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97436 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42959 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12271 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20454 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70843 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28298 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95454 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9325 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83730 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51173 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9022 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1367 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42290 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85162 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1308 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99462 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91118 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19503 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79852 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79610 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79123 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19622 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23663 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19487 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24659 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113766 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19174 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32899 "Found 7 new JSC stress test failures: microbenchmarks/memcpy-wasm-small.js.no-llint, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-collect-continuously, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.wasm-slow-memory, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate-catch.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-eager-jettison, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22634 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20914 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:
```diff
@@ -40,6 +40,12 @@ class BackendDispatcher;
+using CommandResponse = Inspector::Protocol::ErrorStringOr<T>;
+
+template <typename T>
+using CommandCallback = Function<void(CommandResponse<T>)>;
```
out of curiosity, any reason to not use `CompletionHandler` since we always expect it to be called once?

NIT: i feel like maybe `AsyncCommandCallback` (and `AsyncCommandResponse`) would be a bit better to clarify that it's only used by things marked `"async": true` instead of _all_ commands

burg replied:
It has stricter semantics and adopting it everywhere is not always appropriate–and if you mess up it crashes. But now that this refactor is scoped to just WebDriver related code at the moment, it might be workable?

dcrousso commented on Source/JavaScriptCore/inspector/scripts/tests/commands-with-async-attribute.json:
```diff
@@ -27,6 +27,14 @@
     ],
     "commands": [
+        {
+            "name": "executeSQLAsyncNoReturnValues",
```
can we also include the resulting changes to `Source/JavaScriptCore/inspector/scripts/tests/expected/commands-with-async-attribute.json-result`?

burg replied:
The protocol generator needs to be rebaselined prior to this change. I will do that so that the generated code changes associated with this PR are obvious.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
```diff
@@ -203,45 +203,45 @@ class WebAutomationSession final : public API::ObjectImpl<API::Object::Type::Aut
-    Inspector::Protocol::ErrorStringOr<void> deleteAllCookies(const Inspector::Protocol::Automation::BrowsingContextHandle&) override;
-    Inspector::Protocol::ErrorStringOr<Ref<JSON::ArrayOf<Inspector::Protocol::Automation::SessionPermissionData>>> getSessionPermissions() override;
-    Inspector::Protocol::ErrorStringOr<void> setSessionPermissions(Ref<JSON::Array>&&) override;
+    void getBrowsingContexts(Ref<GetBrowsingContextsCallback>&&);
```
why was the `override` removed?

dcrousso commented on Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:
```diff
@@ -40,6 +40,18 @@ class BackendDispatcher;
 typedef String ErrorString;
 
+template<typename T>
+using CommandResult = Inspector::Protocol::ErrorStringOr<T>;
```
i kinda feel like these should be in `Inspector::Protocol` instead of just `Inspector` to match where `ErrorStringOr` and all the other protocol objects are 🤔

burg replied:
I considered this, but it worked well to use CommandResult in non-protocol code like SimulatedInputDispatcher, so I kept this. ErrorStringOr is just Expected<T, String> so we don't even need that template except to interoperate w/ existing code.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -369,10 +369,10 @@ Ref<Inspector::Protocol::Automation::BrowsingContext> WebAutomationSession::buil
 {
     if (pages.isEmpty()) {
-        callback->sendSuccess(WTFMove(contexts));
+        callback({ WTFMove(contexts) });
```
are the `{` `}` actually necessary?

ditto for all the other places that just have a single value in the command result

burg replied:
Good point, maybe not needed anymore.

I waffled on using std::tuple for single-argument return values, but went against it due to the ambiguity of something like `{ { "string" } }` and `{ "string" }` when the "unexpected" type is the same type as the expected type.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -477,11 +477,11 @@ void WebAutomationSession::switchToBrowsingContext(const Inspector::Protocol::Au
         page->setFocus(true);
 
-        callback->sendSuccess();
+        callback({ });
```
i wonder if we should have a `Success` placeholder value or something like that so that it's a little clearer than `{ }`

ideally we'd have commands with `void` results not require any parameters, but not sure if there's a simple way of doing that with `Function` 🤔

burg replied:
I was not able to think up with a cleaner solution in C++ other than `CommandResult<void>`.

`Success` would need to be an associated type or something, since the type of the Expected<T> differs.

Per EWS, in some cases it seems like the initializer list is too hard for GCC to infer the types, so in those cases we need an explicit `expected(T)`.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSessionMacros.h:
```diff
@@ -52,14 +52,26 @@ do { \
 } while (false)
 
-#define ASYNC_FAIL_WITH_PREDEFINED_ERROR(errorName) \
+#define DEPRECATED_ASYNC_FAIL_WITH_PREDEFINED_ERROR(errorName) \
```
do we even need to keep these around anymore if they're not used anywhere?

burg replied:
Oh, good point, I forgot these macros aren't used outside this class.

dcrousso reviewed (approved):
neat!

patrickangle commented on Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_header.py:
```diff
@@ -178,6 +182,39 @@ def _generate_async_handler_declaration_for_command(self, command):
+                parameter_name = 'opt_' + parameter_name
+            returns.append(CppGenerator.cpp_type_for_command_return_declaration(parameter.type, parameter.is_optional))
+
+        # FIXME: Remove legacy_async_callbacks when IndexedDBAgent moves off of it. <rdar://143782962>
```
Nit: Probably better to use the WebKit bug here instead of a radar

patrickangle reviewed (approved):
I’m curious if this change on the WebDriver side is enough to break compatibility for mismatched host+remote sides for WebDriver? It seems like it should be fine given the automation protocol itself hasn’t changed, but curious if we’ve tried?

burg committed:
Web Automation: modernize callback thunks for async protocol commands
https://bugs.webkit.org/show_bug.cgi?id=284470
<rdar://141291944>

Reviewed by Devin Rousso and Patrick Angle.

To avoid code duplication, we need to be able to connect async `WebAutomationSession` / `WebDriverBidiProcessor`
methods to each other. Right now these callbacks are all `BackendDispatcher::CallbackBase` subclasses,
which does not have a standalone constructor. So, we cannot easily pass a lambda that chains completion handlers.

In order to chain callbacks, switch from using `CallbackBase` subclasses to using `Function<>` with proper type parameters.
We can move away from CallbackBase by capturing any needed state in a closure and changing the type of callbacks to mention return types.
We deploy the new type aliases `CommandResponse<T>` and `CommandCallback<T>` to make code more readable, adding the 'Of'
suffix for multiple arguments.

```
namespace Inspector {
    typedef String ErrorString;

    template <typename T>
    using ErrorStringOr = Expected<T, ErrorString>;

    template <typename T>
    using CommandResponse = Inspector::Protocol::ErrorStringOr<T>;

    template <typename T>
    using CommandResponseOf = Inspector::Protocol::ErrorStringOr<std::tuple<T>>;

    template <typename T>
    using CommandCallback = Function<void(CommandResponse<T>)>;

    template <typename T>
    using CommandCallbackOf = Function<void(CommandResponseOf<T>)>;
}
```

These changes are scoped to *not* affect existing Web Inspector protocol generated code. I'd love to remove
uses of `CallbackBase` from the existing uses of 'async' commands entirely. However, doing so is essentially the
same work as refactoring/modernizing IndexedDBAgent. We can do that at a time of our choosing. Fixing this is
tracked by <https://webkit.org/b/288203>.

The remainder of this commit is updating callsites in `WebAutomationSession` to reflect that calling `sendSuccess()` /
`sendFailure()` has been superceded by returning expected or unexpected values of `CommandResult<T>` (aka `ErrorStringOr<T>`).
Also, adapt code to use `CommandCallback<T>` for zero and single parameter async commands, and `CommandCallbackOf<T>` for
async commands with more than one return value.

* Source/JavaScriptCore/inspector/InspectorBackendDispatcher.h:
Introduce new type aliases which make it easier to express the type of an async callback in command implementations.

* Source/JavaScriptCore/inspector/scripts/codegen/cpp_generator_templates.py:
Split out the declaration of the reply thunk class from the command's method declaration.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_header.py:
(CppBackendDispatcherHeaderGenerator._generate_handler_declarations_for_domain):
(CppBackendDispatcherHeaderGenerator._generate_async_response_thunk_declaration_for_command): Added.
(CppBackendDispatcherHeaderGenerator):
(CppBackendDispatcherHeaderGenerator._generate_async_handler_declaration_for_command):
Now that the return types are part of the type signature, enhance `_generate_dispatcher_declaration_for_command`
to compute the type arguments for `Inspector::CommandCallbackOf<T...>`.
Add a new framework setting so that we can continue to generate existing async protocol code
for non-WebDriver domains.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_backend_dispatcher_implementation.py:
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementations_for_domain):
(CppBackendDispatcherImplementationGenerator._generate_async_command_response_thunk_class_for_domain): Added.
(CppBackendDispatcherImplementationGenerator._generate_async_dispatcher_class_for_domain): Deleted.
(CppBackendDispatcherImplementationGenerator._generate_dispatcher_implementation_for_command):
Refactor `_generate_dispatcher_implementation_for_command` so that the code to convert return parameters
is shared between the sync and async command cases. Move the return parameter passing code from the thunk
class to the anonymous lambda function passed as the last argument. Inside the lambda, capture what would
have been stored in the thunk class's member variables so that we can associate the error with the
originating command request (`protocol_requestId`).

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::terminate):
(WebKit::WebAutomationSession::getBrowsingContext):
(WebKit::WebAutomationSession::createBrowsingContext):
(WebKit::WebAutomationSession::switchToBrowsingContext):
(WebKit::WebAutomationSession::setWindowFrameOfBrowsingContext):
(WebKit::WebAutomationSession::waitForNavigationToComplete):
Fix this function so that it explicitly returns after moving `callback` into
the queue of callbacks waiting for navigation. It's not safe to use it anyway,
so simplify the control flow to make it easier to read.

(WebKit::WebAutomationSession::waitForNavigationToCompleteOnPage):
(WebKit::WebAutomationSession::waitForNavigationToCompleteOnFrame):
(WebKit::WebAutomationSession::respondToPendingPageNavigationCallbacksWithTimeout):
(WebKit::WebAutomationSession::respondToPendingFrameNavigationCallbacksWithTimeout):
(WebKit::WebAutomationSession::maximizeWindowOfBrowsingContext):
(WebKit::WebAutomationSession::hideWindowOfBrowsingContext):
(WebKit::WebAutomationSession::willShowJavaScriptDialog):
(WebKit::WebAutomationSession::navigateBrowsingContext):
(WebKit::WebAutomationSession::goBackInBrowsingContext):
(WebKit::WebAutomationSession::goForwardInBrowsingContext):
(WebKit::WebAutomationSession::reloadBrowsingContext):
(WebKit::WebAutomationSession::navigationOccurredForFrame):
(WebKit::WebAutomationSession::documentLoadedForFrame):
(WebKit::WebAutomationSession::inspectorFrontendLoaded):
(WebKit::WebAutomationSession::mouseEventsFlushedForPage):
(WebKit::WebAutomationSession::keyboardEventsFlushedForPage):
(WebKit::WebAutomationSession::wheelEventsFlushedForPage):
(WebKit::WebAutomationSession::willClosePage):
(WebKit::WebAutomationSession::evaluateJavaScriptFunction):
(WebKit::WebAutomationSession::resolveChildFrameHandle):
(WebKit::WebAutomationSession::resolveParentFrameHandle):
(WebKit::WebAutomationSession::computeElementLayout):
(WebKit::WebAutomationSession::getComputedRole):
(WebKit::WebAutomationSession::getComputedLabel):
(WebKit::WebAutomationSession::selectOptionElement):
(WebKit::WebAutomationSession::setFilesForInputFileUpload):
(WebKit::WebAutomationSession::getAllCookies):
(WebKit::WebAutomationSession::deleteSingleCookie):
(WebKit::WebAutomationSession::addSingleCookie):
Add missing move for `callback` to be owned the completion handler.

(WebKit::WebAutomationSession::simulateMouseInteraction):
(WebKit::WebAutomationSession::simulateTouchInteraction):
(WebKit::WebAutomationSession::simulateKeyboardInteraction):
(WebKit::WebAutomationSession::simulateWheelInteraction):
These cases require a bit more care to bridge between `CommandResponse<void>`
and `AutomationCommandError`.

(WebKit::WebAutomationSession::performMouseInteraction):
(WebKit::WebAutomationSession::performKeyboardInteractions):
(WebKit::pressedCharKey):
(WebKit::WebAutomationSession::performInteractionSequence):
(WebKit::WebAutomationSession::cancelInteractionSequence):
(WebKit::WebAutomationSession::takeScreenshot):
Adapt to using `CommandCallback<void>` and `CommandCallback<T>`.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSessionMacros.h:
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::WebAutomationSession::inspectBrowsingContext):
Adapt to using `CommandCallback<void>` and `CommandCallback<T>`.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSessionMacros.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::navigate): Update signature.

Canonical link: https://commits.webkit.org/291284@main
